File: data-pipeline/pipeline/engine/roots.py
"""Bracketed scalar root finding for decreasing functions (Illinois variant of regula falsi).

Both circuit solves are roots of decreasing functions of a log variable: circulating load falls as
the cut coarsens, and product size falls as energy rises. Non-finite values can only occur on the
low side (a vanishing cut traps the pan), and are treated as positive infinity. Written as plain
arithmetic so the TypeScript engine repeats the same iterates.
"""
from __future__ import annotations

import math
from typing import Callable

import numpy as np

from .constants import constant


class RootError(ValueError):
    pass


def _value(f: Callable[[float], float], x: float) -> float:
    try:
        v = f(x)
    except (RootError, FloatingPointError, np.linalg.LinAlgError):
        return math.inf
    return v if math.isfinite(v) else math.inf


def bracket_decreasing(f: Callable[[float], float], x0: float, step: float, lo: float, hi: float) -> tuple[float, float, float, float]:
    """Return ``a < b`` with finite ``f(a) > 0 >= f(b)``."""
    cap = int(constant("numerics.root_max_iterations"))
    x = min(max(x0, lo), hi)
    fx = _value(f, x)
    if fx > 0.0:
        a, fa = x, fx
        for _ in range(cap):
            b = min(hi, a + step)
            fb = _value(f, b)
            if fb <= 0.0:
                break
            if b >= hi:
                raise RootError("no sign change below the upper limit")
            a, fa = b, fb
            step *= 2.0
        else:
            raise RootError("upper bracket search did not converge")
    else:
        b, fb = x, fx
        for _ in range(cap):
            a = max(lo, b - step)
            fa = _value(f, a)
            if fa > 0.0:
                break
            if a <= lo:
                raise RootError("no sign change above the lower limit")
            b, fb = a, fa
            step *= 2.0
        else:
            raise RootError("lower bracket search did not converge")
    for _ in range(cap):
        if math.isfinite(fa):
            return a, b, fa, fb
        m = 0.5 * (a + b)
        fm = _value(f, m)
        if fm > 0.0:
            a, fa = m, fm
        else:
            b, fb = m, fm
    raise RootError("could not find a finite positive bracket end")


def illinois(f: Callable[[float], float], a: float, b: float, fa: float, fb: float) -> float:
    """Root of ``f`` in ``[a, b]`` given finite ``f(a) > 0 >= f(b)``."""
    if fb == 0.0:
        return b
    tol = float(constant("numerics.root_rel_tolerance"))
    side = 0
    c = a
    for _ in range(int(constant("numerics.root_max_iterations"))):
        previous = c
        c = (a * fb - b * fa) / (fb - fa)
        fc = _value(f, c)
        if fc == 0.0 or abs(c - previous) <= tol * max(1.0, abs(c)):
            return c
        if fc < 0.0:
            b, fb = c, fc
            if side == -1:
                fa *= 0.5
            side = -1
        else:
            a, fa = c, fc
            if side == 1:
                fb *= 0.5
            side = 1
    return c


def solve_decreasing(f: Callable[[float], float], x0: float, step: float, lo: float, hi: float) -> float:
    a, b, fa, fb = bracket_decreasing(f, x0, step, lo, hi)
    return illinois(f, a, b, fa, fb)

File: frontend/src/engine/roots.ts
/**
 * Bracketed scalar root finding for decreasing functions, Illinois variant of regula falsi (port of
 * engine/roots.py). The same arithmetic in the same order, so both engines repeat the same iterates.
 */
import { constant } from './constants';
import { SingularMatrixError } from './linalg';

export class RootError extends Error {}

function value(f: (x: number) => number, x: number): number {
  let v: number;
  try {
    v = f(x);
  } catch (error) {
    if (error instanceof RootError || error instanceof SingularMatrixError) return Infinity;
    throw error;
  }
  return Number.isFinite(v) ? v : Infinity;
}

/** Return a < b with finite f(a) > 0 >= f(b). */
export function bracketDecreasing(f: (x: number) => number, x0: number, stepIn: number, lo: number, hi: number): [number, number, number, number] {
  const cap = constant('numerics.root_max_iterations');
  let step = stepIn;
  const x = Math.min(Math.max(x0, lo), hi);
  const fx = value(f, x);
  let a: number; let b: number; let fa: number; let fb: number;
  if (fx > 0.0) {
    a = x; fa = fx; b = a; fb = fa;
    let found = false;
    for (let k = 0; k < cap; k += 1) {
      b = Math.min(hi, a + step);
      fb = value(f, b);
      if (fb <= 0.0) { found = true; break; }
      if (b >= hi) throw new RootError('no sign change below the upper limit');
      a = b; fa = fb;
      step *= 2.0;
    }
    if (!found) throw new RootError('upper bracket search did not converge');
  } else {
    b = x; fb = fx; a = b; fa = fb;
    let found = false;
    for (let k = 0; k < cap; k += 1) {
      a = Math.max(lo, b - step);
      fa = value(f, a);
      if (fa > 0.0) { found = true; break; }
      if (a <= lo) throw new RootError('no sign change above the lower limit');
      b = a; fb = fa;
      step *= 2.0;
    }
    if (!found) throw new RootError('lower bracket search did not converge');
  }
  for (let k = 0; k < cap; k += 1) {
    if (Number.isFinite(fa)) return [a, b, fa, fb];
    const m = 0.5 * (a + b);
    const fm = value(f, m);
    if (fm > 0.0) { a = m; fa = fm; } else { b = m; fb = fm; }
  }
  throw new RootError('could not find a finite positive bracket end');
}

/** Root of f in [a, b] given finite f(a) > 0 >= f(b). */
export function illinois(f: (x: number) => number, aIn: number, bIn: number, faIn: number, fbIn: number): number {
  let a = aIn; let b = bIn; let fa = faIn; let fb = fbIn;
  if (fb === 0.0) return b;
  const tol = constant('numerics.root_rel_tolerance');
  let side = 0;
  let c = a;
  const cap = constant('numerics.root_max_iterations');
  for (let k = 0; k < cap; k += 1) {
    const previous = c;
    c = (a * fb - b * fa) / (fb - fa);
    const fc = value(f, c);
    if (fc === 0.0 || Math.abs(c - previous) <= tol * Math.max(1.0, Math.abs(c))) return c;
    if (fc < 0.0) {
      b = c; fb = fc;
      if (side === -1) fa *= 0.5;
      side = -1;
    } else {
      a = c; fa = fc;
      if (side === 1) fb *= 0.5;
      side = 1;
    }
  }
  return c;
}

export function solveDecreasing(f: (x: number) => number, x0: number, step: number, lo: number, hi: number): number {
  const [a, b, fa, fb] = bracketDecreasing(f, x0, step, lo, hi);
  return illinois(f, a, b, fa, fb);
}

/** Python's round(): halves go to the even integer. */
export function roundHalfEven(x: number): number {
  const floor = Math.floor(x);
  const diff = x - floor;
  if (diff > 0.5) return floor + 1;
  if (diff < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}
